### ISSUE TITLE
clcache.py can not deal with args with double \

### ISSUE THREAD

**Issue by billy0920** [frerich/clcache#31]
when dealing with some args like /Fo"Release\\", clcache.py will faild to regnizing the 2nd quote.

**Comment by jdonald** [comment]
This is the first roadblock I ran into when trying to build our project with clcache. We're using CMake-generated Visual Studio 2013 proj/sln files. The closing doublequote gets misinterpreted as an escaped doublequote, then argument parsing is completely broken after that point.

Oddly the `splitCommandsFile` description says we should handle this case :
```
    # A good test line to split is
    # '"-IC:\\Program files\\Some library" -DX=1 -DVERSION=\\"1.0\\"
    # -I..\\.. -I"..\\..\\lib" -DMYPATH=\\"C:\\Path\\"'
```

(i.e. the C:\Path\ dir has a trailing slash). Yet the logic that follows looks like it's trying to implement UNIX-style doublequote escaping :

```
        if line[i] == '"' and ((i == 0) or (i > 0 and line[i - 1] != '\\')):
            insideQuotes = not insideQuotes
```

So my local fix was to remove that backslash-check, and simply predicate on `if line[i] == '"':` for inverting insideQuotes.

It looks like this code was intended for the rare case of command-line macros that require escaped doublequotes, like `-DVERSION=\\"1.0\\"`, should a project be unfortunate to have those. However, the rules for escaping doublequotes in the Windows command prompt seem to be [complicated](http://stackoverflow.com/questions/7760545/cmd-escape-double-quotes-in-parameter) and nothing like bash, so I don't think this code as written works. I likewise had to remove the `replace('\\"', '"')` step of `extractArgument`.

**Comment by vchigrin** [comment]
@jdonald, Part of splitCommandsFile code was written by me, sorry if it does not work on your project. Handling things like ```-DVERSION=\\"1.0\\"``` really should be kept - I am certainly can remember that this issue were met during building some versions of chromium browser. I am not sure about case
```-DMYPATH=\\"C:\\Path\\"'```, though, we still use clcache on our project and it works fine... 
Could you attach example of command-line that cause wrong behavior of splitCommandsFile ? Is it something like ``` /Fo"Release\"```, as mentioned above?

**Comment by jdonald** [comment]
> Could you attach example of command-line that cause wrong behavior of splitCommandsFile

We have pretty long command lines, but here's the relevant part :
```
u'/Zc:forScope', u'/GR', u'/Fo"CrashReport.dir\\Release\\" /Fd"CrashReport.dir\\Release\\vc120.pdb" /Gd /TP /wd4996 /errorReport:prompt  /w34189 /w34389 /Zm256 G:\\platform\\source\\CrashReport\\CrashReport.cpp']'
...
/Zc:forScope /GR /Fo"CrashReport.dir\Release\" /Fd"CrashReport.dir\Release\vc120.pdb" /Gd /TP /wd4996 /errorReport:prompt  /w34189 /w34389 /Zm256 G:\platform\source\CrashReport\CrashReport.cpp: no source file found
```

>  Is it something like /Fo"Release\", as mentioned above?

Yes, exactly that. Everything after /Fo gets interpreted as one argument and thus it cannot find the source file among other things. Compilation succeeds on fallback, but nothing is cached.

> we still use clcache on our project and it works fine

I suspect that this is common in CMake-generated projects, and @billy0920 may be using CMake as well.

**Comment by webmaster128** [comment]
As @jdonald pointed out, I also don't see a reason for this replacement: `replace('\\"', '"')` in extractArgument.

Why sould `"-DWEBRTC_SVNREVISION=\"Unavailable(issue687)\"` become `-DWEBRTC_SVNREVISION="Unavailable(issue687)` instead of `-DWEBRTC_SVNREVISION=\"Unavailable(issue687)\"`? I think this is a wrong test.

Using `version.cpp` as

```
int main()
{
    auto version = VERSION;
    return 0;
}
```

No. | preprocessor command in cmd | generated source code
-----|-------------------------------|-------------------------------
01 | cl /EHsc /E /D"VERSION=\"1\"" version.cpp | auto version = "1";
01 | cl /EHsc /E /D"VERSION="1"" version.cpp | auto version = 1;
02 | cl /EHsc /E /D"VERSION=1" version.cpp | auto version = 1;
03 | cl /EHsc /E /DVERSION=1 version.cpp | auto version = 1;
04 | cl /EHsc /E /DVERSION="1" version.cpp | auto version = 1;
05 | cl /EHsc /E /DVERSION=\"1\" version.cpp | auto version = "1";
06 | cl /EHsc /E "/DVERSION=\"1\"" version.cpp | auto version = "1";
07 | cl /EHsc /E "/DVERSION=\"1\"" version.cpp | auto version = "1";
08 | cl /EHsc /E "/DVERSION="1"" version.cpp | auto version = 1;
09 | cl /EHsc /E /DVERSION=\\\\"1\\\\" version.cpp | auto version = \1\;

Comparing 07 with 04 shows that the replacement changes meaning of the argument.

The double backslash version in 09 looks entirely broken to me and causes a compile bug when run directly on cl.exe.

**Comment by webmaster128** [comment]
Okay, this is my idea on that issue: #47

**Comment by frerich** [comment]
This ought to be fixed now after merging pull request #47
